src/runtime/pipeline/pipeline_executor.h

```diff
@@ -36,25 +43,89 @@ namespace runtime {
  *
  *  This executor can be accessed by various language via TVM runtime PackedFunc API.
  */
-class TVM_DLL PipelineRuntime : public ModuleNode {
+class TVM_DLL PipelineExecutor : public ModuleNode {
  public:
   /*!
    * \Return the type key of the executor.
    */
-  const char* type_key() const final { return "PipelineRuntime"; }
+  const char* type_key() const final { return "PipelineExecutor"; }
   /*!
    * \brief Initialize the pipeline executor with module array and json text.
    * \param modules The module list used for building pipeline.
    * \param pipeline_json The configuration of modules dependencies.
    */
-  void Init(const Array<tvm::runtime::Module>& modules, const std::string& pipeline_json);
+  void Init(const Array<Module>& modules, const std::string& pipeline_json);
   /*!
    * \brief Give frontends an access to packed functions.
    * \param name The name of the function.
    * \param sptr_to_self The pointer to the module node.
    * \return The corresponding packed function.
    */
   virtual PackedFunc GetFunction(const std::string& name, const ObjectPtr<Object>& sptr_to_self);
+
+  /*!
+   * \brief Get the number of outputs.
+   *
+   * \return The number of outputs.
+   */
+  int NumOutputs() const { return num_outputs_; }
+
+ private:
+  /*!\brief The class used to execute pipeline logic*/
+  PipelineFunction pipeline_function_;
+  /*!\brief The Dependency information of each graph runtime module of pipeline.*/
+  PipelineConfig pipeline_config_;
+  /*!\brief The Module information that can get used to create graph runtime.*/
+  ModuleConfig mod_config_;
+  /*!\birief How many outputs are in this pipeline executor.*/
+  size_t num_outputs_ = 0;
+  /*!\brief Json loader.*/
+  void Load(dmlc::JSONReader* reader) {
+    reader->BeginArray();
+    while (reader->NextArrayItem()) {
+      std::string key;
+      reader->BeginObject();
+      int mod_idx = 0;
+      std::string lib_name;
+      std::string json_name;
+      std::string params_name;
+      std::string dev;
+      OutputMap output;
+      while (reader->NextObjectItem(&key)) {
+        if (key == "mod_idx") {
+          reader->Read(&mod_idx);
+        }
+        if (key == "lib_name") {
+          reader->Read(&lib_name);
+        }
+
+        if (key == "json_name") {
+          reader->Read(&json_name);
+        }
+
+        if (key == "params_name") {
+          reader->Read(&params_name);
+        }
+
+        if (key == "dev") {
+          reader->Read(&dev);
+        }
+
+        if (key == "output") {
+          reader->Read(&output);
+        }
+      }
+      // Check if mod_idx is read successfully.
+      ICHECK(mod_idx > 0);
+      // Check if the output is read successfully.
+      ICHECK(!output.Empty());
+      pipeline_config_.Insert(mod_idx, output);
+      // Check if there is lib, json and params information.
+      if (!lib_name.empty() && !json_name.empty() && !params_name.empty()) {
```

Comment:
throw an error otherwise?

Reply:
these 3 file used to load graph module, if the graph module already get set by the function PipelineExecutor::Init, then these 3 files  being empty is ok, so here just check the empty and not throw a error.

Reply:
Not a good design. No ones will know this behavior by reading this function.
If the graph module is already initialized, why we need to call this function again?

Reply:
add 'load_module' parameter into 'Load', once the 'modules' parameter in 'PipelineExecutor::Init' is empty, this variable get set into true, then do lib_name/json_name/params_name check and load from disk.